File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
class Solution {
    public int solution(int[][] triangle) {
        int answer = 0;
        int [][] memory = new int[501][501];
        memory[0][0] = triangle[0][0];
        
        for(int i=0; i<triangle.length-1; i++){
            for(int j=0; j<=i; j++){
                int res = memory[i][j] + triangle[i+1][j];
                if(memory[i+1][j]<res) memory[i+1][j] = res;
                
                if(j!=i+1){
                    int res1 = memory[i][j] + triangle[i+1][j+1];
                    if(memory[i+1][j+1]<res) memory[i+1][j+1] = res1;
                }
            }
        }
        
        for(int i=0; i<=triangle.length; i++){
            for(int j=0; j<=i; j++){
               answer = memory[i][j] > answer ? memory[i][j] : answer ;
            }
        }
        
        return answer;
    }
}

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

int solution(vector<vector<int>> triangle) {
    int N = triangle.size();
    vector<vector<int>> sum(N, vector<int>(N,0));
   
    sum[0][0]=triangle[0][0];
    for(int i=1; i<triangle.size(); i++){
        for(int j=0; j<=i; j++){
            if(j==0){ //좌측 끝에 붙어있는 경우 
                sum[i][j]=sum[i-1][j]+triangle[i][j];
            }else{ //그 외 
                sum[i][j]=max(sum[i-1][j],sum[i-1][j-1])+triangle[i][j];
            }
        }
    }
    return *max_element(sum.at(N-1).begin(),sum.at(N-1).end());
}